fwidgen/src/main.rs

```diff
@@ -0,0 +1,133 @@
+use anyhow::{anyhow, Context, Result};
+use clap::Parser;
+use hubtools::RawHubrisArchive;
+use sha3::{Digest, Sha3_256};
+use std::{
+    str,
+    ops::Range,
+};
+
+// may already be in lpc55_support?
+pub const LPC55_FLASH_PAGE_SIZE: usize = 512;
+
+#[derive(Parser, Debug)]
+struct Args {
+    /// Hubris archive
+    #[clap(env = "HUBEDIT_ARCHIVE")]
+    archive: String,
+}
+
+#[derive(Debug)]
+enum Chip {
+    Lpc55,
+    Stm32,
+}
+
+impl TryFrom<&RawHubrisArchive> for Chip {
+    type Error = anyhow::Error;
+
+    fn try_from(archive: &RawHubrisArchive) -> Result<Self> {
+        let manifest = archive.extract_file("app.toml")?;
+        let manifest: toml::Value = toml::from_str(
+            str::from_utf8(&manifest).context("manifest bytes to UTF8")?,
+        ).context("manifest UTF8 to TOML")?;
+
+        let chip = manifest
+           .as_table()
+           .ok_or(anyhow!("manifest isn't a table"))?
+           .get("chip")
+           .ok_or(anyhow!("no key \"chip\" in manifest"))?
+           .as_str()
+           .ok_or(anyhow!("value for key \"chip\" isn't a string"))?;
+
+        if chip.contains("lpc55") {
+            Ok(Chip::Lpc55)
+        } else if chip.contains("stm32") {
+            Ok(Chip::Stm32)
+        } else {
+            Err(anyhow!("Unsupported chip: {}", chip))
+        }
+    }
+}
```

Comment:
This is almost completely reimplementing `is_lpc55` in hubtools. (https://github.com/oxidecomputer/hubtools/blob/main/hubtools/src/lib.rs#L832-L852) . Is there any way we could avoid that duplication?

Reply:
I considered adding a new `is_stm32` implementation to hubtools so we had parity w/ `is_lpc55`. That's kinda clunky though so I implemented this enum w/o disturbing the existing `is_lpc55` function. My ideal end state would be to remove `is_lpc55` and replace it w/ something like this `enum Chip` implementing `TryFrom<RawHubrisArchive>` .... but since that is a breaking API change so I put it off.

If you think it should be done as part of this PR I'm happy to move `enum Chip` to `hubtools`. Removing `is_lpc55` may cause breakage so maybe mark it as deprecated (compiler warning)?

Reply:
I like your proposal for future work and think it's fine to do it separately. We could always re implement `is_lpc55` with the enum underneath for backwards compatibility.

Reply:
that's a great idea, added it to https://github.com/oxidecomputer/hubtools/issues/39

Reply:
taking your `LGTM` here as a soft approval, merging and moving on to #39 etc